### PR TITLE
refactor: Simplify the runtime selection

### PR DESCRIPTION
**Pull request by Marwes**
The runtime selection has been very complicated and filled with duplicated code. This doesnt't do away with all of this complication but it does at least encapsulate it a bit better and opens a path for using a non-tcp/unix based connections (say for mocking)

### CODE REVIEW

File: src/aio_async_std.rs [hunk 1]
```diff
@@ -1,151 +1,156 @@
...
         buf: &[u8],
-    ) -> std::task::Poll<Result<usize, tokio::io::Error>> {
-        async_std::io::Write::poll_write(Pin::new(&mut self.0), cx, buf)
+    ) -> Poll<io::Result<usize>> {
+        match &mut *self {
+            AsyncStd::Tcp(r) => Pin::new(r).poll_write(cx, buf),
+            #[cfg(feature = "tokio-tls-comp")]
+            AsyncStd::TcpTls(r) => Pin::new(r).poll_write(cx, buf),
+            #[cfg(unix)]
+            #[cfg(feature = "tokio-comp")]
```
Comment (weihanglo) [inline src/aio_async_std.rs:98]:
```suggestion
```

File: src/aio_async_std.rs [hunk 2]
```diff
@@ -1,151 +1,156 @@
...
         mut self: Pin<&mut Self>,
-        cx: &mut core::task::Context,
+        cx: &mut task::Context,
         buf: &[u8],
-    ) -> std::task::Poll<Result<usize, tokio::io::Error>> {
-        async_std::io::Write::poll_write(Pin::new(&mut self.0), cx, buf)
+    ) -> Poll<io::Result<usize>> {
+        match &mut *self {
+            AsyncStd::Tcp(r) => Pin::new(r).poll_write(cx, buf),
+            #[cfg(feature = "tokio-tls-comp")]
```
Comment (weihanglo) [inline src/aio_async_std.rs:95]:
```suggestion
            #[cfg(feature = "async-std-tls-comp")]
```

File: src/aio_async_std.rs [hunk 3]
```diff
@@ -1,151 +1,156 @@
...
 
-    fn poll_flush(
-        mut self: Pin<&mut Self>,
-        cx: &mut core::task::Context,
-    ) -> std::task::Poll<Result<(), tokio::io::Error>> {
-        async_std::io::Write::poll_flush(Pin::new(&mut self.0), cx)
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<io::Result<()>> {
+        match &mut *self {
+            AsyncStd::Tcp(r) => Pin::new(r).poll_flush(cx),
+            #[cfg(feature = "tokio-tls-comp")]
```
Comment (weihanglo) [inline src/aio_async_std.rs:106]:
```suggestion
            #[cfg(feature = "async-std-tls-comp")]
```

File: src/aio_async_std.rs [hunk 4]
```diff
@@ -1,151 +1,156 @@
...
 
-    fn poll_shutdown(
-        mut self: Pin<&mut Self>,
-        cx: &mut core::task::Context,
-    ) -> std::task::Poll<Result<(), tokio::io::Error>> {
-        async_std::io::Write::poll_close(Pin::new(&mut self.0), cx)
+    fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<io::Result<()>> {
+        match &mut *self {
+            AsyncStd::Tcp(r) => Pin::new(r).poll_shutdown(cx),
+            #[cfg(feature = "tokio-tls-comp")]
```
Comment (weihanglo) [inline src/aio_async_std.rs:116]:
```suggestion
            #[cfg(feature = "async-std-tls-comp")]
```

File: src/aio_async_std.rs [hunk 5]
```diff
@@ -1,151 +1,156 @@
...
         mut self: Pin<&mut Self>,
-        cx: &mut core::task::Context,
+        cx: &mut task::Context,
         buf: &mut [u8],
-    ) -> std::task::Poll<Result<usize, tokio::io::Error>> {
-        async_std::io::Read::poll_read(Pin::new(&mut self.0), cx, buf)
+    ) -> Poll<io::Result<usize>> {
+        match &mut *self {
+            AsyncStd::Tcp(r) => Pin::new(r).poll_read(cx, buf),
+            #[cfg(feature = "tokio-tls-comp")]
```
Comment (weihanglo) [inline src/aio_async_std.rs:132]:
```suggestion
            #[cfg(feature = "async-std-tls-comp")]
```

File: src/client.rs [hunk 6]
```diff
@@ -55,70 +58,54 @@ impl Client {
...
+        use crate::aio::RedisRuntime;
+        Ok(
+            crate::aio::connect::<crate::aio_async_std::AsyncStd>(&self.connection_info)
+                .await?
+                .map(RedisRuntime::boxed),
+        )
     }
 
     /// Returns an async connection from the client.
     #[cfg(all(feature = "aio", feature = "tokio-rt-core"))]
```
Comment (weihanglo) [inline src/client.rs:100]: Should either require `tokio-rt-core` or `async-std-comp` but not always `tokio-rt-core`.

```suggestion
    #[cfg(any(feature = "tokio-rt-core", all(not(feature = "tokio-comp"), feature = "async-std-comp")))]
```

File: src/client.rs [hunk 7]
```diff
@@ -166,7 +150,8 @@ impl Client {
         crate::aio::MultiplexedConnection,
```
Comment (weihanglo) [inline src/client.rs:150]: `fn create_multiplexed_tokio_connection()` config should become `#[cfg(feature = "tokio-rt-core")]`

File: src/client.rs [hunk 8]
```diff
@@ -206,10 +192,88 @@ impl Client {
...
+        T: crate::aio::RedisRuntime,
+    {
+        let (connection, driver) = self
+            .create_multiplexed_async_connection_inner::<T>()
+            .await?;
+        T::spawn(driver);
+        Ok(connection)
+    }
+
+    #[cfg(all(feature = "aio", feature = "tokio-rt-core"))]
```
Comment (weihanglo) [inline src/client.rs:212]:
```suggestion
    #[cfg(any(feature = "tokio-rt-core", feature = "async-std-comp"))]
```

File: src/client.rs [hunk 9]
```diff
@@ -134,24 +121,21 @@ impl Client {
...
+        self.get_multiplexed_async_connection_inner::<crate::aio::tokio_aio::Tokio>()
+            .await
     }
 
     /// Returns an async multiplexed connection from the client.
     ///
     /// A multiplexed connection can be cloned, allowing requests to be be sent concurrently
     /// on the same underlying connection (tcp/unix socket).
-    ///
     /// This requires the `tokio-rt-core` feature as it uses the default tokio executor.
```
Comment (weihanglo) [inline src/client.rs:132]:
```suggestion
    /// This requires the `async-std-comp` feature as it uses the default async-std executor.
```

File: src/client.rs [hunk 10]
```diff
@@ -206,10 +192,88 @@ impl Client {
     /// [multiplexed-connection]: aio/struct.MultiplexedConnection.html
     #[cfg(feature = "connection-manager")]
     pub async fn get_tokio_connection_manager(&self) -> RedisResult<crate::aio::ConnectionManager> {
-        Ok(crate::aio::ConnectionManager::new(self.connection_info.clone()).await?)
+        Ok(crate::aio::ConnectionManager::new(self.clone()).await?)
+    }
+
+    #[cfg(all(feature = "aio", feature = "tokio-rt-core"))]
```
Comment (weihanglo) [inline src/client.rs:198]:
```suggestion
    #[cfg(any(feature = "tokio-rt-core", feature = "async-std-comp"))]
```

File: src/client.rs [hunk 11]
```diff
@@ -55,70 +58,54 @@ impl Client {
...
-            } else {
-                self.get_multiplexed_async_std_connection().await
-            }
-        }
-
-        #[cfg(all(not(feature = "tokio-comp"), not(feature = "async-std-comp")))]
-        {
-            compile_error!("tokio-comp or async-std-comp features required for aio feature")
+        match self.get_runtime() {
+            #[cfg(feature = "tokio-comp")]
```
Comment (weihanglo) [inline src/client.rs:105]:
```suggestion
            #[cfg(feature = "tokio-rt-core")]
```

**Comment by weihanglo** [comment]
Can also delete this feature config.

https://github.com/mitsuhiko/redis-rs/blob/4503f33e98c84c71fa66d4d78dcf195054862776/src/client.rs#L119

**Comment by weihanglo** [comment]
@Marwes 
Thanks for accepting my suggestions 😎 

In case you miss them, three little feature configs need some changes: https://github.com/mitsuhiko/redis-rs/pull/360#discussion_r460603528, https://github.com/mitsuhiko/redis-rs/pull/360#discussion_r460603093, https://github.com/mitsuhiko/redis-rs/pull/360#issuecomment-664078499. Feel free to discuss 😊

**Comment by weihanglo** [comment]
Not to offend someone. I know that not everyone has spare time to review code and I am here just trying to help. IMO redis-rs is a awesome and stable project but a bit stagnant these days. To make some progresses on recently pull requests, I suggest a further discussion to expand the maintainership. Or maybe we can at least try to merge some reasonable pull requests such as this great refactor PR. 😎 

I have no idea who is in charge of redis-rs currently. Forgive my boldly pinging @mitsuhiko @Marwes @badboy. Ping @Terkwood  seems terkwood is also interesting in this project.

Thanks again for all your contributions! 😊

**Comment by Marwes** [comment]
Since I mostly contribute to redis-rs for work (where we rely redis quite a bit) this got a bit stuck as I were away for 5 weeks over summer. Still got some catching up after getting back to work but I will get back and finish this soon-ish at least.

File: src/client.rs [hunk 12]
```diff
@@ -160,13 +156,15 @@ impl Client {
     /// A multiplexed connection can be cloned, allowing requests to be be sent concurrently
     /// on the same underlying connection (tcp/unix socket).
     #[cfg(feature = "tokio-comp")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "tokio-comp")))]
```
Comment (weihanglo) [inline src/client.rs:159]:
```suggestion
    #[cfg(feature = "tokio-rt-core")]
    #[cfg_attr(docsrs, doc(cfg(feature = "tokio-rt-core")))]
```

File: src/client.rs [hunk 13]
```diff
@@ -51,107 +57,97 @@ impl Client {
...
     /// Returns an async multiplexed connection from the client.
     ///
     /// A multiplexed connection can be cloned, allowing requests to be be sent concurrently
     /// on the same underlying connection (tcp/unix socket).
-    ///
-    /// This requires the `tokio-rt-core` feature as it uses the default tokio executor.
-    #[cfg(feature = "tokio-rt-core")]
     #[cfg(feature = "tokio-comp")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "tokio-rt-core")))]
+    #[cfg_attr(docsrs, doc(cfg(feature = "tokio-comp")))]
```
Comment (weihanglo) [inline src/client.rs:132]: Multiplexed connection needs spawn fn (provided by tokio/rt-core).

```suggestion
    #[cfg(feature = "tokio-rt-core")]
    #[cfg_attr(docsrs, doc(cfg(feature = "tokio-rt-core")))]
```

**Review by weihanglo — COMMENTED** [review]
Even better 🎉 
Only some misconfigurations and typos need fixes

**Review by weihanglo — APPROVED** [review]

**Pull request merged**